### PR TITLE
Fix rolling update for StatefulSet integration

### LINKED ISSUE #3690: Rolling update does not work for StatefulSet integration
/kind bug
/kind feature

**What happened**:

When doing a rolling update of StatefulSet it gets stuck after stopping the first pod. 

**What you expected to happen**:

The rolling update of STS is supported as this is a common operation for serving workloads.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create the STS as here: 

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: nginx-statefulset
  labels:
    app: nginx
    kueue.x-k8s.io/queue-name: user-queue
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: registry.k8s.io/nginx-slim:0.26
          ports:
            - containerPort: 80
          resources:
            requests:
              cpu: "100m"
  serviceName: "nginx"
```

2. trigger STS rolling update with:

```sh
kubectl set image statefulset/nginx-statefulset nginx=registry.k8s.io/nginx-slim:0.27
```
issue: The update gets stuck, and the pods are:
```
> k get pods      
NAME                  READY   STATUS      RESTARTS   AGE
nginx-statefulset-0   1/1     Running     0          3m33s
nginx-statefulset-1   1/1     Running     0          3m30s
nginx-statefulset-2   0/1     Completed   0          3m29s
```

**Anything else we need to know?**:

After deleting the finalizer from `nginx-statefulset-2` pod manually, the STS update is still stuck, the pods:

```
> k get pods                            
NAME                  READY   STATUS            RESTARTS   AGE
nginx-statefulset-0   1/1     Running           0          5m51s
nginx-statefulset-1   1/1     Running           0          5m48s
nginx-statefulset-2   0/1     SchedulingGated   0          3s
```

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fix rolling update for StatefulSet integration

#### Which issue(s) this PR fixes:

Fixes #3690

#### Does this PR introduce a user-facing change?

```release-note
Fixed rolling update for StatefulSet integration
```

### CODE REVIEW

File: test/e2e/singlecluster/statefulset_test.go [hunk 1]
```diff
@@ -130,6 +130,46 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
 			util.ExpectObjectToBeDeletedWithTimeout(ctx, k8sClient, createdWorkload, false, util.LongTimeout)
 		})
 
+		ginkgo.It("should allow to update pod spec", func() {
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:133]:
```suggestion
		ginkgo.It("should allow to update the PodTemplate in StatefulSet", func() {
```

File: test/e2e/singlecluster/statefulset_test.go [hunk 2]
```diff
@@ -130,6 +130,46 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
...
+			createdStatefulSet := &appsv1.StatefulSet{}
+			ginkgo.By("Waiting for replicas is ready", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(statefulSet), createdStatefulSet)).
+						To(gomega.Succeed())
+					g.Expect(createdStatefulSet.Status.ReadyReplicas).To(gomega.Equal(int32(3)))
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("Update pod spec", func() {
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:153]:
```suggestion
			ginkgo.By("Update the PodTemplate in StatefulSet", func() {
```

File: test/e2e/singlecluster/statefulset_test.go [hunk 3]
```diff
@@ -130,6 +130,46 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
...
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(statefulSet), createdStatefulSet)).To(gomega.Succeed())
+					g.Expect(createdStatefulSet.Spec.Template.Spec.Containers).Should(gomega.HaveLen(1))
+					createdStatefulSet.Spec.Template.Spec.Containers[0].Image = util.E2eTestSleepImage
+					g.Expect(k8sClient.Update(ctx, createdStatefulSet)).To(gomega.Succeed())
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			pods := &corev1.PodList{}
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:162]: wrap this in a ginkgo.By("Await for the Pods to be replaced with the new template")

File: test/e2e/singlecluster/statefulset_test.go [hunk 4]
```diff
@@ -130,6 +130,46 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
...
+			gomega.Eventually(func(g gomega.Gomega) {
+				g.Expect(k8sClient.List(ctx, pods, client.InNamespace(ns.Name), client.MatchingLabels(createdStatefulSet.Spec.Selector.MatchLabels))).To(gomega.Succeed())
+				g.Expect(pods.Items).To(gomega.HaveLen(3))
+				for _, p := range pods.Items {
+					g.Expect(createdStatefulSet.Spec.Template.Spec.Containers).Should(gomega.HaveLen(1))
+					g.Expect(p.Spec.Containers[0].Image).To(gomega.Equal(util.E2eTestSleepImage))
+				}
+			}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+		})
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:171]: Please also add a set to "Await for the StatefulSet to have all replicas ready again", and check the `status.ReadyReplicas=3`.

File: test/integration/controller/jobs/pod/pod_controller_test.go [hunk 5]
```diff
@@ -1403,6 +1403,108 @@ var _ = ginkgo.Describe("Pod controller", ginkgo.Ordered, ginkgo.ContinueOnFailu
...
+
+				ginkgo.By("finalize pod2", func() {
+					gomega.Eventually(func(g gomega.Gomega) {
+						g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(pod2), createdPod2)).Should(gomega.Succeed())
+						createdPod2.Finalizers = nil
+						g.Expect(k8sClient.Update(ctx, createdPod2)).Should(gomega.Succeed())
+					}, util.Timeout, util.Interval).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("check pod2 is finalize", func() {
```
Comment (mimowo) [inline test/integration/controller/jobs/pod/pod_controller_test.go:1485]:
```suggestion
				ginkgo.By("await for pod2 to be deleted", func() {
```
?

File: pkg/controller/jobs/pod/pod_controller_test.go [hunk 6]
```diff
@@ -2640,6 +2640,123 @@ func TestReconciler(t *testing.T) {
 			},
 			workloadCmpOpts: defaultWorkloadCmpOpts,
 		},
+		"reclaimable pods not updated for serving pod group": {
```
Comment (mimowo) [inline pkg/controller/jobs/pod/pod_controller_test.go:2643]:
```suggestion
		"reclaimablePods field is not updated for a serving pod group": {
```

**Review by mimowo — COMMENTED** [review]
LGTM, just nits from me. I think we could consider a dedicated KEP for serving workloads, where we could mention the new annotation, but I wouldn't like to block the PR on it before 0.10, as this is an important fix. 

cc @tenzen-y

File: test/e2e/singlecluster/statefulset_test.go (hunk 3 above)
Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go:162]: Oops. Missed that.

Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go]: Done

File: test/integration/controller/jobs/pod/pod_controller_test.go (hunk 5 above)
Comment (mbobrovskyi) [reply test/integration/controller/jobs/pod/pod_controller_test.go]: Done

File: pkg/controller/jobs/pod/pod_controller_test.go (hunk 6 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/pod/pod_controller_test.go]: Done

File: test/e2e/singlecluster/statefulset_test.go (hunk 2 above)
Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go]: Done

**Comment by mimowo** [comment]
> LGTM, just nits from me. I think we could consider a dedicated KEP for serving workloads, where we could mention the new annotation (and the fast-admission annotation added before), but I wouldn't like to block the PR on it before 0.10, as this is an important fix.
> 
> cc @tenzen-y

I synced with @mbobrovskyi and he will create a KEP for serving workloads - starting from the StatefulSet support. We will then extend the KEP with LeaderWorkerSet when we have a prototype. But I think it would be great to merge this for 0.10 anyway.

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
/hold
For another look from @tenzen-y 
Please also add the KEP in a separate PR with the two annotations we introduced for the support: "fast-admission" and "serving".

**Comment by mimowo** [comment]
/lgtm
/hold 
for the approval of the KEP update https://github.com/kubernetes-sigs/kueue/pull/3692
cc @tenzen-y

**Comment by mimowo** [comment]
/approve
/hold cancel
Moving forward along with the KEP update, see also: https://github.com/kubernetes-sigs/kueue/pull/3692#issuecomment-2517332000

File: test/util/e2e.go [hunk 7]
```diff
@@ -26,6 +26,8 @@ import (
 )
 
 const (
+	// E2eTestSleepImageOld is the image used for testing rolling update.
+	E2eTestSleepImageOld = "gcr.io/k8s-staging-perf-tests/sleep:v0.0.3@sha256:00ae8e01dd4439edfb7eb9f1960ac28eba16e952956320cce7f2ac08e3446e6b"
```
Comment (tenzen-y) [inline test/util/e2e.go:30]: Sorry for the late. This is just a question. Why do we need to use this outdated image?

**Pull request merged**